crates/matrix-sdk-ui/src/timeline/controller/aggregations.rs

```diff
@@ -0,0 +1,429 @@
+    PollResponse {
+        /// Sender of the poll's response.
+        sender: OwnedUserId,
+        /// Timestamp at which the response has beens ent.
+        timestamp: MilliSecondsSinceUnixEpoch,
+        /// All the answers to the poll sent by the sender.
+        answers: Vec<String>,
+    },
+
+    /// This is the marker of the end of a poll.
+    PollEnd {
+        /// Timestamp at which the poll ends, i.e. all the responses with a
+        /// timestamp prior to this one should be taken into account
+        /// (and all the responses with a timestamp after this one
+        /// should be dropped).
+        end_date: MilliSecondsSinceUnixEpoch,
+    },
+
+    /// This is a reaction to another event.
+    Reaction {
+        /// The reaction "key" displayed by the client, often an emoji.
+        key: String,
+        /// Sender of the reaction.
+        sender: OwnedUserId,
+        /// Timestamp at which the reaction has been sent.
+        timestamp: MilliSecondsSinceUnixEpoch,
+        /// The send status of the reaction this is, with handles to abort it if
+        /// we can, etc.
+        reaction_status: ReactionStatus,
+    },
+}
+
+/// An aggregation is an event related to another event (for instance a
+/// reaction, a poll's response, etc.).
+///
+/// It can be either a local or a remote echo.
+#[derive(Clone, Debug)]
+pub(crate) struct Aggregation {
+    /// The kind of aggregation this represents.
+    pub kind: AggregationKind,
+
+    /// The own timeline identifier for a reaction.
+    ///
+    /// It will be a transaction id when the aggregation is still a local echo,
+    /// and it will transition into an event id when the aggregation is a
+    /// remote echo (i.e. has been received in a sync response):
+    pub own_id: TimelineEventItemId,
+}
+
+/// Get the poll state from a given [`TimelineItemContent`].
+fn poll_state_from_item(
+    content: &mut TimelineItemContent,
+) -> Result<&mut PollState, AggregationError> {
+    match content {
+        TimelineItemContent::Poll(poll_state) => Ok(poll_state),
+        _ => Err(AggregationError::InvalidType {
+            expected: "a poll".to_owned(),
+            actual: content.debug_string().to_owned(),
+        }),
+    }
+}
+
+impl Aggregation {
+    /// Create a new [`Aggregation`].
+    pub fn new(own_id: TimelineEventItemId, kind: AggregationKind) -> Self {
+        Self { kind, own_id }
+    }
+
+    /// Apply an aggregation in-place to a given [`TimelineItemContent`].
+    ///
+    /// In case of success, returns a boolean indicating whether applying the
+    /// aggregation caused a change in the content. If the aggregation could be
+    /// applied, but it didn't cause any change in the passed
+    /// [`TimelineItemContent`], `Ok(false)` will be returned.
+    ///
+    /// In case of error, returns an error detailing why the aggregation
+    /// couldn't be applied.
+    pub fn apply(&self, content: &mut TimelineItemContent) -> Result<bool, AggregationError> {
```

Comment:
As discussed irl, returning an enum instead of the bool here would make it clearer, I think.